Rewrite RHEL 9 and CS9 image definitions using the new framework

achilleas-k opened the pull request:
This is a rewrite of the `distro/rhel9/` package to use the new image definition framework that was so far used only for Fedora.
Some changes are not final but will probably be addressed in a follow-up rather than here.

A rewrite of `distro/rhel8/` is also planned, but will be done separately.  That should require less work after everything that went into this one.

I've described and provided context for all changes in individual commit messages.

Remaining tasks:
- [x] image installer
- [x] tar
- [x] change edge tests to check for `sysroot=ro` on RHEL 9 and CS9
- [ ] ~~edge simplified installer~~ (will be handled separately)

### Updates
1. The edge simplified installer will be done in a follow-up.
https://github.com/osbuild/osbuild-composer/pull/3120#issuecomment-1323593242

2. Changes to installer definitions.
https://github.com/osbuild/osbuild-composer/pull/3120#issuecomment-1324116599

3. Status of sysroot readonly for RHEL 9 and CS9 edge images.
https://github.com/osbuild/osbuild-composer/pull/3120#issuecomment-1324407557

achilleas-k commented:
@lavocatt Is there a way to get manifest-db to run image info on all of these and check for changes?  This is one of the cases where that would be perfect since all the manifests have changed slightly but the image info should be identical.

lavocatt commented:
> @lavocatt Is there a way to get manifest-db to run image info on all of these and check for changes? This is one of the cases where that would be perfect since all the manifests have changed slightly but the image info should be identical.

Not yet, because it fetches the last manifests on top of composer's main. I could hack something for you to specify a branch name for composer in a variable on the trigger workflow ?

thozza commented on internal/runner/rhel.go:
```diff
@@ -13,7 +13,8 @@ func (r *RHEL) String() string {
 	packages := []string{
-		"glibc", // ldconfig
+		"glibc",           // ldconfig
+		"platform-python", // osbuild
```
I'm really sure this will work due to https://bugzilla.redhat.com/show_bug.cgi?id=2121775

achilleas-k replied:
I realised this was missing because I tried to build all the manifests that changed and they were failing immediately.  Adding this was enough for RHEL 9.  I guess if we find that we need `python3` for RHEL 8 we can add that too.

thozza replied:
Interesting. I though (based on the bug report) that osbuild stages are actually not able to use `platform-python` due to hardcoding the binary path in the shebang. So installing it should not help much 🤔 I didn't investigate this more at all so I may be all wrong...

achilleas-k replied:
The [RHEL 8.x runners](https://github.com/osbuild/osbuild/blob/f34bee944bdfc590cf25159521ee66c515d0cf10/runners/org.osbuild.rhel82#L1) use `platform-python`, so that's all we need in the build root.

I think the bug report is a case of removing the dependency from the host and not being able to bootstrap the build root.

thozza replied:
Thanks for the additional info. This is very helpful!

However I'm afraid that it is not completely true, since `python36` gets pulled into the buildroot pipeline and it is installed there. So while you are right that the bug is caused by the situation that you've described, AFAICT osbuild stages still use the `python3` installed in the buildroot, because they are executed in a separate process without an explicit interpreter path provided in the [`subprocess.run()` call](https://github.com/osbuild/osbuild/blob/f34bee944bdfc590cf25159521ee66c515d0cf10/runners/org.osbuild.rhel82#L76-L78).

gicmo replied:
Our current RHEL 8 (and thus 9) runner also has an implicit dependency on `python3.6`:

https://github.com/osbuild/osbuild/blob/f34bee944bdfc590cf25159521ee66c515d0cf10/runners/org.osbuild.rhel82#L54-L63

thozza replied:
That's exactly the point that I'm trying to make... Therefore the buildroot packages should explicitly list `python36` and not rely on it being pulled in as a dependency of something else that is installed in the buildroot...

gicmo replied:
See also https://bugzilla.redhat.com/show_bug.cgi?id=2121775

achilleas-k replied:
So do we add `python36` to the runner deps for RHEL to be explicit?

thozza replied:
So it looks like it is pulled into the DNF transaction by some other package, so specifying it in the build packages will have basically no real effect.

Personally I suggest to add it there to make the dependency explicit and prevent any unexpected issues due to `python36` not being installed. It should be there until the rhel8 runner is "fixed" to depend only on `platform-python`.

Including a comment about the whole situation explaining two pythons being listed would be probably also good, but I'm pretty sure you would add something like this anyway 😉

achilleas-k replied:
Done!

thozza commented on internal/distro/rhel9/gce.go:
```diff
@@ -0,0 +1,290 @@
+		basePartitionTables: defaultBasePartitionTables,
+	}
+
+	// default AMI (EC2 BYOS) images config
```
The comment is wrong "GCE" vs. "AMI"

achilleas-k replied:
Oh, it was deleted in a followup commit after the copy/paste to multiple files.  I've cleared it from that commit as well.

thozza commented on internal/distro/rhel9/imagetype.go:
```diff
@@ -327,6 +390,65 @@ func (t *imageType) Manifest(customizations *blueprint.Customizations,
+	// create a manifest object and instantiate it with the computed packageSetChains
+	manifest, err := t.initializeManifest(&bp, options, globalRepos, packageSets, nil, 0)
+	if err != nil {
+		// TODO: handle manifest initialization errors more gracefully, we
+		// refuse to initialize manifests with invalid config.
+		return nil
```
you should at least print the error here (as done in #3132)

achilleas-k replied:
Done!

thozza commented on internal/distro/rhel9/imagetype.go:
```diff
@@ -273,13 +287,62 @@ func (t *imageType) PartitionType() string {
+		return nil, err
+	}
+
+	// TODO: let image types specify valid workloads, rather than
+	// always assume Custom.
+	w := &workload.Custom{
+		BaseWorkload: workload.BaseWorkload{
+			Repos: packageSets[blueprintPkgsKey].Repositories,
+		},
+		Packages: bp.GetPackagesEx(false),
+	}
```
Note: We will actually have to implement a way to merge workloads. The reason is that if we define SAP as a workload, instead of an image type, then we would still have to have a way for user to customize image with SAP workload with potentially additional packages, enabled services, etc... Also the complication is that we now depsolve packages in chain where the workload packages are always depsolved as a 2nd transaction, which may not be desirable for "standard" workloads like SAP, but really only for packages provided by the user in BP. Probably not that relevant for your rework, but this will have to be figured out before we use Workloads for anything else than BP customizations.

teg replied:
My thinking with workloads was that we would not expose the same options in SAP as Custom, but only those options that make sense in that setting. I imagine that for an SAP image we would not support adding extra packages, systems services etc, maybe we would not support any options at all, or just a few SAP specific ones. The benefit of the named workloads would be that we know the use case so can be more restrictive.

thozza replied:
hmm, I see. It is definitely an option. If that would be the case, then things could stay as they are. However it would also mean regressing in our current functionality, because we support customization of any image type, even those that are focused on specific workload.

So as a result, I expected that we want to allow building custom images of any type. I can imagine that the user may need to install some additional packages and services also into predefined-workload-specific images for monitoring, inspection of the system, etc. E.g. also we do forward logs to Splunk, the customer's security team may require a special software to be installed on the OS for CVE and threat assessment.

With the SAP image as an example, it is preconfigured for SAP, but does not come with SAP software. Maybe someone would want to build custom SAP images with the actual SAP software installed from custom repos 🤔 I'm not actually sure, just describing situations in which it would make sense to allow (package and service) customizations. The workload-specific customizations are definitely an interesting idea and IMO would make sense.

teg replied:
I agree, I also don't know for certain. It could be that we'll need a lot (all?) the customisations for some specific workloads, but I think we should have the discussion on a case-by-case basis on what makes sense, to limit our test exposure etc.

Do we already expose SAP images with customisations anywhere? I thought they were only built internally?

thozza replied:
> Do we already expose SAP images with customisations anywhere? I thought they were only built internally?

The image type is filtered out by default via Weldr API. With a configuration change and adding a few new repo definitions for SAP-specific packages would definitely make it possible to build it on-premise. Nevertheless this is not really documented anywhere and would require knowing what to do. So we can IMO safely consider the SAP image in general not being buildable on-premise.

For Clous API (including the internal service), any customizations can be already passed when building the SAP image (as well as any other image). AFAIK it is not really used by anyone nor by any pipeline. My point was that it is possible, not that it is used.

teg replied:
Thanks for the context!

thozza reviewed (commented):
Just submitting comments from my partial review... This is huge so more to come 😆 (note that I was adding comments as I was going through separate commits)

thozza commented on internal/manifest/os.go:
```diff
@@ -176,14 +193,30 @@ func (p *OS) getPackageSetChain() []rpmmd.PackageSet {
 	if p.SElinux != "" {
-		packages = append(packages, "policycoreutils")
-		packages = append(packages, fmt.Sprintf("selinux-policy-%s", p.SElinux))
+		packages = append(packages, "policycoreutils", fmt.Sprintf("selinux-policy-%s", p.SElinux))
```
Note that this is actually wrong (https://github.com/osbuild/osbuild-composer/issues/2954). We should take the policy specified in the image configuration into account. I think that those two values should be eventually merged and that's actually the plan (as to move away from osbuild-specific data structures). Just pointing out that this currently uses hardcoded value.

achilleas-k replied:
ah good to know, thanks.  I was just merging the two lines here, but good that you noticed so we can fix it now if it's not too big a change.

thozza replied:
I think that it is fine to use this as is for the first rework, but we should make sure to not introduce any regressions in RHEL images before the development window for 9.2 and 8.8 closes...

achilleas-k commented on tools/test-case-generators/format-request-map.json:
```diff
@@ -717,6 +734,92 @@
       }
     },
     "overrides": {
+      "rhel-91": {
```
They will, I was just too lazy to copy-paste the overrides for all since they use the same pipelines, it should be fine.  When RHEL 8 gets the feature, we can remove the override and make it part of the standard `with-customizations` variant.

thozza replied:
I see. It would be probably much easier if we could use wildcards in the distro name in `format-request-map.json`...

thozza commented on internal/image/live.go:
```diff
@@ -79,6 +79,9 @@ func (img *LiveImage) InstantiateManifest(m *manifest.Manifest,
 	case platform.FORMAT_GCE:
+		// NOTE(akoutsou): temporary workaround; filename required for GCP
+		// TODO: define internal raw filename on image type
+		imagePipeline.Filename = "disk.raw"
```
Yeah, the filename really matters. Requirements are defined in https://cloud.google.com/compute/docs/import/import-existing-image#requirements_for_the_image_file. You probably plan it anyway, but I suggest to merge this fix and the regenerated GCE manifests with the previous commit that changes the pipeline used for GCE images and the commit which regenerated GCE manifests (It contained a commit message that the filename does not matter, which would be confusing (because it does).

achilleas-k replied:
yeah thanks.  I knew GCP had strict requirements for the archive but forgot that the internal filename was one of them until I saw the test failing.

Rebased and squashed.

achilleas-k commented:
> Remaining tasks:
> 
> * [x]  image installer
> * [ ]  edge simplified installer
> * [x]  tar
> * [x]  change edge tests to check for `sysroot=ro` on RHEL 9 and CS9

The edge simplified installer will be done in a follow-up.

achilleas-k commented:
### Ready for review!

#### Some notes on the most recent changes.

I've made some changes to the new installer definitions that were added in #3077.

I added options so that each distro can configure things a little differently when it comes to the kickstart file and the squashfs compression. 

For the kickstart file: Fedora uses the default kickstart location (interactive-defaults) but in RHEL so far we put the kickstart file in the root of the ISO at `/osbuld.ks`.  For now, I decided we should maintain the old behaviour for RHEL 9 and CS9 and make changes later.  This would require changing tests and possibly docs, so the change should be more isolated and purposeful.

For the squashfs compression: I couldn't get the RHEL 9.2 ISO to boot with lz4 compression, which is used in Fedora.  So far we used xz compression in RHEL and CS, so again I'm keeping the old behaviour.  If we need to change it, we can figure out what the issue is and make the change.

achilleas-k commented:
> * [x]  change edge tests to check for `sysroot=ro` on RHEL 9 and CS9

This was reverted.  Since anaconda doesn't set `sysroot` to read only in RHEL 9, our raw image would be inconsistent with installer deployments.  We should make the change in RHEL when anaconda does the same.

teg commented on internal/manifest/tar.go:
```diff
@@ -10,24 +10,27 @@ type Tar struct {
 	}
-	if inputPipeline.manifest != m {
-		panic("tree pipeline from different manifest")
-	}
```
Should this be replaced by anything? Again, not worth a rebase.

achilleas-k replied:
Yes it's unfortunate that we lost the sanity checks but we could figure out a new way to do it.

teg commented on internal/distro/rhel9/images.go:
```diff
@@ -346,6 +346,9 @@ func imageInstallerImage(workload workload.Workload,
 	img.SquashfsCompression = "xz"
 
+	// put the kickstart file in the root of the iso
+	img.ISORootKickstart = true
```
What's the ideal? I remember there was a discussion with the installer team, but forgot the outcome. If not for compat, where should we be putting the .ks?

achilleas-k replied:
For this PR, I'm making sure to not make any changes to image definitions, but we should think about this and decide what we want.

gicmo replied:
https://hackmd.io/hwMbx1CXQRKvHXGW8yQTMg

teg commented on tools/test-case-generators/format-request-map.json:
```diff
@@ -490,6 +490,12 @@
         "name": "image-installer-users",
+        "packages": [
+          {
+            "name": "podman",
```
Why?

achilleas-k replied:
1. Why not?
2. Good to have a package in there to see that it goes in the right place (payload packages), because the package set handling for the installer is a little different than other image types.  The Fedora installer in main right now ignores user/blueprint packages.

teg reviewed (dismissed):
Love all of this!
I threw in a bunch of comments/questions, but regardless of the answers, please address anything in follow-ups.

achilleas-k commented on internal/distro/rhel9/distro.go:
```diff
@@ -640,41 +620,61 @@ func (t *imageType) checkOptions(customizations *blueprint.Customizations, optio
+	return NewRHEL92()
+}
+
+func newDistro(name string, major, minor int) distro.Distro {
```
Yes.  Good point!

achilleas-k commented on internal/distro/rhel9/distro.go:
```diff
@@ -237,6 +237,7 @@ func newDistro(name string, major, minor int) distro.Distro {
 		bootType: distro.LegacyBootType,
 	}
 
+	qcow2ImgType := mkQcow2ImgType(rd)
```
No, it's right.  The problem is that the local `qcow2ImgType` here shadows the package global one which doesn't have an image config.  Very error prone.

achilleas-k commented on internal/distro/rhel9/distro.go:
```diff
@@ -327,10 +326,20 @@ func newDistro(name string, major, minor int) distro.Distro {
 				ImageFormat: platform.FORMAT_RAW,
 			},
-			BIOS:       false,
+			BIOS:       true,
```
Diff artefact.  The actual change is splitting one platform into two:

From
```go
	x86_64.addImageTypes(
		&platform.X86{
			BasePlatform: platform.BasePlatform{
				ImageFormat: platform.FORMAT_RAW,
			},
			BIOS:       false,
			UEFIVendor: rd.vendor,
		},
		edgeRawImgType,
		edgeSimplifiedInstallerImgType,
	)

```

To
```go
	x86_64.addImageTypes(
		&platform.X86{
			BasePlatform: platform.BasePlatform{
				ImageFormat: platform.FORMAT_RAW,
			},
			BIOS:       true,
			UEFIVendor: rd.vendor,
		},
		edgeRawImgType,
	)

	x86_64.addImageTypes(
		&platform.X86{
			BasePlatform: platform.BasePlatform{
				ImageFormat: platform.FORMAT_RAW,
			},
			BIOS:       false,
			UEFIVendor: rd.vendor,
		},
		edgeSimplifiedInstallerImgType,
	)
```

The raw image got moved to a new platform that supprts BIOS.

gicmo replied:
Tripped over this too, it is fine! ;)

nullr0ute commented on internal/platform/ppc64le.go:
```diff
@@ -0,0 +1,33 @@
+	if p.BIOS {
+		return "powerpc-ieee1275"
+	}
+	return ""
```
BIOS is an x86 term, the POWER firmware is OpenFirmware based, so depends on the context.

achilleas-k replied:
We generally (and sloppily) use BIOS to mean "non-UEFI".

gicmo reviewed (dismissed):
❤️ 
Went through it commit by commit. Found one minor thing, but I am not telling because it is not a big deal. ;)

thozza commented on internal/distro/rhel9/images.go:
```diff
@@ -111,6 +111,9 @@ func osCustomizations(
 			},
 		)
 	}
+	if imageConfig.Keyboard != nil && imageConfig.Keyboard.X11Keymap != nil {
+		osc.X11Layouts = imageConfig.Keyboard.X11Keymap.Layouts
+	}
```
This should be moved to the block above:
```go
if keyboard != nil {
	osc.Keyboard = keyboard
} else if imageConfig.Keyboard != nil {
	osc.Keyboard = &imageConfig.Keyboard.Keymap
}
```
The whole config (keymap and x11layouts) are currently applied as a whole - https://github.com/osbuild/osbuild-composer/blob/3443484502b448fd00c5596ab546c190be781518/internal/distro/rhel9/pipelines.go#L420-L424

Lastly, the `X11Keymap.Layouts` tend to be related to the value set in `Keyboard`.

achilleas-k replied:
The way it's done now (see block below) allows for a customization to change the base `Keymap` but still maintain the `X11Keymap.Layouts` if the image config defines them.
Indeed, they are related, but since we don't support customizing the ``X11Keymap` (yet?), I wasn't sure whether an external customization for the base `Keymap` should override it or not.

```go
	if keyboard != nil {
		osc.Keyboard = keyboard
	} else if imageConfig.Keyboard != nil {
		osc.Keyboard = &imageConfig.Keyboard.Keymap
	}
	if imageConfig.Keyboard != nil && imageConfig.Keyboard.X11Keymap != nil {
		osc.X11KeymapLayouts = imageConfig.Keyboard.X11Keymap.Layouts
	}
```

achilleas-k replied:
I guess keeping the image config can lead to weird issues where you have an X11KeymapLayout that's a whole different thing from the base Keymap, so it makes sense to override completely.  I'll change it.

thozza replied:
Yeah, my point was that the image config values (all of them) related to keyboard settings should be thrown away if the user specified anything in the Blueprint. At least that's how it works now and IMO it makes sense. The reason is that the X11Keymap layout tends to be related to the Keymap value, so overriding just one of them with user-provided value and applying the other as it was set in image config does not IMO make sense.

achilleas-k commented:
Rebased and fixed merge conflicts

thozza reviewed (commented):
Ended up in a half, submitting what I found so far...

teg commented on internal/distro/rhel9/qcow2.go:
```diff
@@ -158,3 +143,31 @@ func openstackCommonPackageSet(t *imageType) rpmmd.PackageSet {
+	ic := &distro.ImageConfig{
+		DefaultTarget: common.StringToPtr("multi-user.target"),
+	}
+	if d.isRHEL() {
```
We would have to make it conditional on subscription options being enabled.

gicmo commented on internal/distro/rhel9/azure.go:
```diff
@@ -30,7 +24,7 @@ var (
 		bootable:            true,
 		defaultSize:         4 * common.GibiByte,
-		pipelines:           vhdPipelines(false),
+		image:               liveImage,
```
I think this commit also switches the azure image type over to the new style, but the commit message doesn't say so. Just in case you are changing something in the PR anyway.

achilleas-k replied:
oooh that's weird.  I might try to separate them actually.

achilleas-k replied:
Separated.

achilleas-k commented on internal/runner/centos.go:
```diff
@@ -12,12 +12,24 @@ func (c *CentOS) String() string {
 	}
+	if c.Version < 9 {
+		packages = append(packages,
+			// The RHEL 8 runner (which is also used for CS8) in osbuild runs
+			// with platform-python but explicitly symlinks python 3.6 to
+			// /etc/alternatives (which in turn is the target for
+			// /usr/bin/python3) for the stages.
+			// https://github.com/osbuild/osbuild/blob/ea8261cad6c5c606c00c0f2824c3f483c01a0cc9/runners/org.osbuild.rhel82#L61
+			// Install python36 explicitly for CS8.
```
Yeah it seems we need it for both.

achilleas-k commented on internal/runner/rhel.go:
```diff
@@ -13,12 +13,23 @@ func (r *RHEL) String() string {
 	}
+	if p.Major < 9 {
+		packages = append(packages,
+			// The RHEL 8 runner in osbuild runs with platform-python but
+			// explicitly symlinks python 3.6 to /etc/alternatives (which in turn
+			// is the target for /usr/bin/python3) for the stages.
+			// https://github.com/osbuild/osbuild/blob/ea8261cad6c5c606c00c0f2824c3f483c01a0cc9/runners/org.osbuild.rhel82#L61
+			// Install python36 explicitly for RHEL 8.
```
Yeah it seems we need it for both.

achilleas-k commented on internal/distro/rhel9/qcow2.go:
```diff
@@ -158,3 +143,31 @@ func openstackCommonPackageSet(t *imageType) rpmmd.PackageSet {
+}
+
+func mkQcow2ImgType(d distribution) imageType {
+	it := qcow2ImgType
```
Yes, this came up here: https://github.com/osbuild/osbuild-composer/pull/3120#discussion_r1031490462
I wasn't going to change it if no other changes were necessary, but I'll do it now since it can lead to problems.

achilleas-k commented on internal/distro/rhel9/ami.go:
```diff
@@ -491,38 +403,101 @@ func rhelEc2SapPackageSet(t *imageType) rpmmd.PackageSet {
+	it := amiImgTypeX86_64
+	ic := defaultAMIImageConfigX86_64(osVersion, rhsm)
+	if rhsm && common.VersionLessThan(osVersion, "9.1") {
+		// Don't disable RHSM redhat.repo management on the AMI
+		// image, which is BYOS and does not use RHUI for content.
+		// Otherwise subscribing the system manually after booting
+		// it would result in empty redhat.repo. Without RHUI, such
+		// system would have no way to get Red Hat content, but
+		// enable the repo management manually, which would be very
+		// confusing.
```
Right.  Done.

achilleas-k commented on internal/manifest/os.go:
```diff
@@ -45,6 +45,7 @@ type OSCustomizations struct {
 	GPGKeyFiles      []string
 	Language         string
 	Keyboard         *string
+	X11Layouts       []string
```
Done.

achilleas-k commented:
Addressed comments.  Force push diff: https://github.com/osbuild/osbuild-composer/compare/91aa9469eb2dab8b710db504bb52e79d71e0a58d..0b1fca9dcebdf04ed5e22e4997cd1a2307746e69

thozza reviewed (commented):
Overall great work 👏 

I found the RHSM configuration rework in EC2 images to be messy, but functionally it works AFAICT. I don't consider any of my comments to be blockers, I'm fine with a follow-up given the size of this PR...

achilleas-k commented on internal/image/live.go:
```diff
@@ -47,8 +48,15 @@ func (img *LiveImage) InstantiateManifest(m *manifest.Manifest,
+			artifact = xzPipeline.Export()
+		default:
+			imagePipeline.Filename = img.Filename
+			artifact = imagePipeline.Export()
```
This appears outdated but the same applies to the latest version.  Good idea, I'll panic.

achilleas-k commented on test/data/manifests/rhel_90-x86_64-ec2-boot.json:
```diff
@@ -2360,7 +2140,7 @@
                 "origin": "org.osbuild.pipeline",
                 "references": {
                   "name:image": {
-                    "file": "image.raw"
```
Yes, seeing how everything works now, it might be worth having a way to specify the base filename separately from the final artifact filename in cases where an image is compressed or tarred.

achilleas-k commented on internal/distro/rhel9/ami.go:
```diff
@@ -491,38 +403,101 @@ func rhelEc2SapPackageSet(t *imageType) rpmmd.PackageSet {
+func mkEc2ImgTypeX86_64(osVersion string, rhsm bool) imageType {
+	it := ec2ImgTypeX86_64
+	ic := defaultEc2ImageConfigX86_64(osVersion, rhsm)
+	if rhsm && common.VersionLessThan(osVersion, "9.1") {
+		// Disable RHSM redhat.repo management
+		rhsmConf := ic.RHSMConfig[distro.RHSMConfigNoSubscription]
+		rhsmConf.SubMan.Rhsm = &osbuild.SubManConfigRHSMSection{ManageRepos: common.BoolToPtr(false)}
+		ic.RHSMConfig[distro.RHSMConfigNoSubscription] = rhsmConf
+	}
```
Good catch.  Removed!

achilleas-k commented on internal/distro/rhel9/ami.go:
```diff
@@ -491,38 +403,101 @@ func rhelEc2SapPackageSet(t *imageType) rpmmd.PackageSet {
+	return rhsm.InheritFrom(ic)
+}
+
+func appendEC2Dracut(ic *distro.ImageConfig) *distro.ImageConfig {
```
Done!

thozza reviewed (dismissed):
LGTM (to the extent that Github was able to actually show, but I went commit by commit before...)

thozza commented on internal/distro/rhel9/ami.go:
```diff
@@ -491,38 +403,95 @@ func rhelEc2SapPackageSet(t *imageType) rpmmd.PackageSet {
+func mkAMIImgTypeX86_64(osVersion string, rhsm bool) imageType {
+	it := amiImgTypeX86_64
+	ic := defaultAMIImageConfigX86_64(osVersion, rhsm)
+	if rhsm && common.VersionLessThan(osVersion, "9.1") {
+		ic = appendRHSM(ic)
```
This should have been removed as well, since it is already done in `defaultAMIImageConfig()` called from `defaultAMIImageConfigX86_64()`. Also the configuration is applied unconditionally to AMIs regardless of the minor version so the condition in `defaultAMIImageConfig()` is actually the correct one. 

Nevertheless this "bug" has no effect on resulting functionality.

achilleas-k replied:
I'll clean this up in a followup.

achilleas-k commented:
Found a bug in the installers.  The UEFI kernel command line arguments weren't being set properly for the kickstart file.
Fixed.

gicmo reviewed (approved):
Reviewed only the recent changes, looking good.
